python-api/tests/integrationtests2/baseclass.py

```diff
@@ -0,0 +1,118 @@
+"""
+Contains the Base class for integration tests.
+"""
+
+import sys
+import os
+import logging
+
+sys.path.insert(0, os.path.abspath(os.path.join(__file__, "../../../")))
+from gstswitch.server import Server
+from gstswitch.helpers import TestSources  # PreviewSinks
+from gstswitch.controller import Controller
+from gi.repository import GLib
+
+# PATH = os.getenv("HOME") + '/gst/stage/bin/'
+PATH = '../tools/'
+
+
+class IntegrationTestbase(object):
+    """Base class for integration tests."""
+
+    # Tests are not allowed to have an __init__ method
+    # pylint: disable=attribute-defined-outside-init
+    def setup_method(self, _):
+        """Set up called automatically before every test_XXXX method."""
+        self.log = logging.getLogger()
+        self.log.setLevel(logging.DEBUG)
+        logging.basicConfig(
+            format="%(filename)s:%(lineno)d (%(funcName)s): %(message)s")
+
+        self.serv = None
+        self._sources = None
+        self.controller = None
+
+    def setup_server(self, record_file=False):
+        """Set up a gst-switch server for testing."""
+        assert self.serv is None
+
+        self.log.info("setting up Server")
+        self.serv = Server(path=PATH, video_format="debug", record_file=record_file)
+
+        self.log.info("running Server")
+        self.serv.run()
+        assert self.serv.pid
+
+        self.log.info("waiting for Server to open Controller-Port")
+        self.serv.wait_for_output('tcp:host=0.0.0.0,port=5000')
+
+        self.log.info("setting up TestSources")
+        self._sources = TestSources(
+            video_port=self.serv.video_port,
+            audio_port=self.serv.audio_port)
+
+    def setup_controller(self):
+        """Create Controller object and call setup_controller."""
+        self.log.info("setting up Controller")
+        self.controller = Controller()
+
+        self.log.info("connecting Controller to Server")
+        self.controller.establish_connection()
+        self.serv.wait_for_output('registered: ')
+
+        assert self.controller.connection is not None
+
+    def new_test_video(self):
+        self.serv.wait_for_output('0.0.0.0:3000')
```

Comment:
Hm I can see your concerns but I did extensive Statistics on a multitude of Systems and with the current output of the Server I have a 100.0% Success.

Technically Speaking - as long as the Server prints() after it opened the port, the Kernel should ensure that these operations are sequential, so the print always comes after the actual network operation.

For Example that's why I moved the Print-Line in 55910db